localstack/services/visitors.py

```diff
@@ -0,0 +1,145 @@
+import importlib
+import logging
+from functools import singledispatchmethod
+from typing import Any, Dict, List, Optional, Protocol, TypedDict, runtime_checkable
+
+from moto.core.utils import BackendDict
+
+from localstack.services.stores import AccountRegionBundle
+
+LOG = logging.getLogger(__name__)
+
+
+class StateVisitor:
+    def visit(self, state_container: Any):
+        """
+        Visit (=do something with) a given state container. A state container can be anything that holds service state.
+        An AccountRegionBundle, a moto BackendDict, or a directory containing assets.
+        """
+        raise NotImplementedError
+
+
+@runtime_checkable
+class StateVisitable(Protocol):
+    def accept(self, visitor: StateVisitor):
+        """
+        Accept a StateVisitor. The implementing method should call visit not necessarily on itself, but can also call
+        the visit method on the state container it holds. The common case is calling visit on the stores of a provider.
+        :param visitor: the StateVisitor
+        """
+
+
+class ServiceBackend(TypedDict, total=False):
+    """Wrapper of the possible type of backends that a service can use."""
+
+    localstack: AccountRegionBundle | None
+    moto: BackendDict | Dict | None
+
+
+class ServiceBackendCollectorVisitor:
+    """Implementation of StateVisitor meant to collect the backends that a given service use to hold its state."""
+
+    def __init__(self) -> None:
+        self.store: AccountRegionBundle | None = None
+        self.backend_dict: BackendDict | Dict | None = None
```

Comment:
i would put the annotations into the class:

```suggestion

    store: AccountRegionBundle | None
    backend_dict: BackendDict | Dict | None

    def __init__(self) -> None:
        self.store = None
        self.backend_dict = None
```